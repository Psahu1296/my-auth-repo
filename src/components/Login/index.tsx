import "./Login.css";
import React, { useReducer } from "react";
import { ActionType, CounterAction, LoginProps, StateInterface } from "./LoginTypes";

// username: "mor_2314",
// password: "83r5^_",
const URL = "https://fakestoreapi.com/auth/login";

const initialState: StateInterface = {
  username: "",
  password: "",
  error: "",
};

export const userReducer = (initialState:StateInterface, action:CounterAction): StateInterface => {
  switch(action.type) {
    case ActionType.USERNAME:
      return {...initialState, username: action.payload}
    case ActionType.PASSWORD:
      return {...initialState, password: action.payload}
    case ActionType.ERROR:
      return {...initialState, error: action.payload}
    default:
      return initialState
  }
} 
const Login = ({ loginFlagHandler }: LoginProps) => {
  const [state, dispatch] = useReducer(userReducer, initialState);

  const changeHandler = (
    e: React.ChangeEvent<HTMLInputElement>,
    type: string
  ) => {
    if (state.error) {
      dispatch({type: ActionType.ERROR, payload: ""});
    }
    if (type === "user") {
      dispatch({type: ActionType.USERNAME , payload: e.target.value});
    } else if (type === "pass") {
      dispatch({type: ActionType.PASSWORD , payload: e.target.value});
    }
  };

  const loginCallHandler = async () => {
    try {
      if (!state.username || !state.password) {
        dispatch({type: ActionType.ERROR, payload: "Your username or password must be incorrect"});
        return;
      }
      if (state.username && state.password) {
        const userData = {
          username: state.username,
          password: state.password,
        };
        const response = await fetch(URL, {
          method: "POST",
          headers: {
            Accept: "application/json, text/plain, */*",
            "Content-type": "application/json; charset=UTF-8",
          },
          body: JSON.stringify(userData),
        });
        const data = await response.json();
        sessionStorage.setItem("isAuthorized", data.token);
        loginFlagHandler(true);
      }
    } catch (err) {
      console.log("Error=>>", err);
      loginFlagHandler(false);
      dispatch({type: ActionType.ERROR, payload: "Something went wrong!"});
    }
  };
  return (
    <div className="overlay" data-testid="login-page">
      <form
        data-testid="login-form"
        className="wrapper"
        onSubmit={e => {
          e.preventDefault();
          loginCallHandler();
        }}
      >
        <h1 data-testid="title" className="title">
          Sign In
        </h1>
        <div className="input_wrapper">
          <label id="username" data-testid="user-label" htmlFor="username">
            Username
          </label>
          <input
            data-testid="text-input"
            type="text"
            placeholder="Username"
            className="input"
            value={state.username}
            onChange={e => changeHandler(e, "user")}
          />
        </div>
        <div className="input_wrapper">
          <label id="password" data-testid="password-label" htmlFor="password">
            Password
          </label>
          <input
            data-testid="password-input"
            type="password"
            placeholder="Password"
            className="input"
            value={state.password}
            onChange={e => changeHandler(e, "pass")}
          />
        </div>
        {state.error && <span data-testid="error-msg" className="error-text">
          *{state.error}
        </span>}
        <button data-testid="submit-btn" type="submit" className="sign-in-btn">
          Sign in
        </button>
      </form>
    </div>
  );
};

export default Login;
