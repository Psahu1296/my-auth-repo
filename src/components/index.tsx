export { default as Login } from "./Login"
export { default as Product } from "./Product"
export { default as Products } from "./Products"