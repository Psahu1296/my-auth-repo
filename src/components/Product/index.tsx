import { ProductParams } from "./ProductsTypes";
import "../Products/Products.css"

const Product = (params: ProductParams) => {
  return (
    <div data-testid="product-id" className="wrapper">
      <span data-testid="product-title" className="title">
        Title: {params.title}
      </span>
      <span data-testid="product-price" >RS :{params.price}</span>
      <span data-testid="product-category">Category: {params.category}</span>
      {/* <span >Description: {params.description}</span> */}
      <img className="product-image" src={params.image} alt="product" />
    </div>
  );
};

export default Product;
