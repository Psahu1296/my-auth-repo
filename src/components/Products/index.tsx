import React, { useEffect, useState } from "react";
import "./Products.css";
import { ProductParams } from "@/components/Product/ProductsTypes";
import Product from "@/components/Product";

const Products = () => {
  const [products, setProducts] = useState<ProductParams[]>([]);

  useEffect(() => {
    fetch("https://fakestoreapi.com/products")
      .then(res => res.json())
      .then(json => setProducts(json));
  }, []);

  return (
    <div className="main-wrapper" data-testid="products-page">
      {products.length > 0 ?products.map((product: ProductParams) => (
        <Product key={product.id} {...product} />
      )): <div className="" data-testid="no-product">No Products found</div>}
    </div>
  );
};

export default Products;

