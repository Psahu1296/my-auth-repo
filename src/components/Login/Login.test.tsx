import { fireEvent, render, screen, waitFor } from "@testing-library/react";

import { Login } from "@/components";

const loginFlagHandler = jest.fn();

describe("<Login/>", () => {
  it("Component should render properly", () => {
    render(<Login loginFlagHandler={loginFlagHandler} />);
    expect(screen.getByTestId("login-page")).toMatchSnapshot();
    expect(screen.getByTestId("title")).toHaveTextContent(/Sign In/);
    expect(screen.getByTestId("user-label")).toHaveTextContent(/Username/);
    expect(screen.getByTestId("text-input")).toBeInTheDocument();
    expect(screen.getByTestId("password-label")).toHaveTextContent(/Password/);
    expect(screen.getByTestId("password-input")).toBeInTheDocument();
    expect(screen.getByTestId("submit-btn")).toHaveTextContent(/sign in/i);
    expect(screen.queryByTestId("error-msg")).not.toBeInTheDocument();
  });

  it("Login successfully", async () => {
    render(<Login loginFlagHandler={loginFlagHandler} />);
    const loginForm = screen.getByTestId("login-form");
    fireEvent.change(screen.getByTestId("text-input"), {
      target: { value: "name" },
    });
    fireEvent.change(screen.getByTestId("password-input"), {
      target: { value: "abcde" },
    });
    fireEvent.submit(loginForm);
    await waitFor(() => {
      expect(loginFlagHandler).toHaveBeenCalled();
      expect(screen.queryByTestId("error-msg")).not.toBeInTheDocument();
    });
  });

  it("Login failed", async () => {
    render(<Login loginFlagHandler={loginFlagHandler} />);
    const loginForm = screen.getByTestId("login-form");
    fireEvent.submit(loginForm);
    await waitFor(() => {
      expect(screen.getByTestId("error-msg")).toBeInTheDocument();
    });
  });
});
