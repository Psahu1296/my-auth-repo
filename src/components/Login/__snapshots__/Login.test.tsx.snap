// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Login/> Component should render properly 1`] = `
<div
  class="overlay"
  data-testid="login-page"
>
  <form
    class="wrapper"
    data-testid="login-form"
  >
    <h1
      class="title"
      data-testid="title"
    >
      Sign In
    </h1>
    <div
      class="input_wrapper"
    >
      <label
        data-testid="user-label"
        for="username"
        id="username"
      >
        Username
      </label>
      <input
        class="input"
        data-testid="text-input"
        placeholder="Username"
        type="text"
        value=""
      />
    </div>
    <div
      class="input_wrapper"
    >
      <label
        data-testid="password-label"
        for="password"
        id="password"
      >
        Password
      </label>
      <input
        class="input"
        data-testid="password-input"
        placeholder="Password"
        type="password"
        value=""
      />
    </div>
    <button
      class="sign-in-btn"
      data-testid="submit-btn"
      type="submit"
    >
      Sign in
    </button>
  </form>
</div>
`;
